src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs

```diff
@@ -16,23 +19,52 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
 
             public int[] ToArray()
             {
-                int[] array = new int[_end - _start];
-                Fill(array, _start);
+                int start = _start;
+                int[] array = new int[_end - start];
+                Fill(array, start);
                 return array;
             }
 
             public List<int> ToList()
             {
-                List<int> list = new List<int>(_end - _start);
-                Fill(SetCountAndGetSpan(list, _end - _start), _start);
+                int start = _start;
+                int end = _end;
+                List<int> list = new List<int>(end - start);
+                Fill(SetCountAndGetSpan(list, end - start), start);
                 return list;
             }
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void Fill(Span<int> destination, int start)
             {
-                for (int i = 0; i < destination.Length; i++, value++)
+                // The improvements in the compiler now allow us to write the loop as Vector256.
+                // Manually unrolling to Vector128x2 for SSE2/NEON is no longer profitable.
+                if (destination.Length < Vector256<int>.Count)
                 {
-                    destination[i] = value;
+                    goto Scalar;
+                }
+
+                int stride = Vector256<int>.Count;
+
+                Vector256<int> init = Vector256.Create(
+                    (ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7 });
+                Vector256<int> value = Vector256.Create(start) + init;
+                Vector256<int> increment = Vector256.Create(stride);
+
+                do
+                {
+                    value.StoreUnsafe(ref MemoryMarshal.GetReference(destination));
```

Comment:
Get the `ref` once before the loop and use that `ref` throughout.

Reply:
Fixed. The codegen did not change though.